File: scripts/oop-practice-sources/cpp/smartphone.cpp
#include <iostream>
#include <string>
#include <algorithm>

class Smartphone {
private:
    std::string model;
    int battery;

public:
    explicit Smartphone(const std::string& model) : model(model), battery(20) {}

    void call() {
        battery = std::max(0, battery - 5);
        std::cout << "Звонок с " << model << "... Заряд: " << battery << "%" << std::endl;
    }

    void charge() {
        battery = std::min(100, battery + 30);
        std::cout << "Зарядка " << model << "... Заряд: " << battery << "%" << std::endl;
    }

    void show_status() const {
        std::cout << "Смартфон " << model << ": заряд " << battery << "%" << std::endl;
    }
};

int main() {
    Smartphone phone("Pixel");
    phone.show_status();
    phone.call();
    phone.charge();
    phone.show_status();
    return 0;
}

File: scripts/oop-practice-sources/lua/smartphone.lua
local Smartphone = {}

function Smartphone:new(model)
    local obj = setmetatable({model = model, _battery = 20}, {__index = self})
    return obj
end

function Smartphone:call()
    self._battery = math.max(0, self._battery - 5)
    print("Звонок с " .. self.model .. "... Заряд: " .. self._battery .. "%")
end

function Smartphone:charge()
    self._battery = math.min(100, self._battery + 30)
    print("Зарядка " .. self.model .. "... Заряд: " .. self._battery .. "%")
end

function Smartphone:show_status()
    print("Смартфон " .. self.model .. ": заряд " .. self._battery .. "%")
end

local phone = Smartphone:new("Pixel")
phone:show_status()
phone:call()
phone:charge()
phone:show_status()
